tests/ui/consts/const-eval/heap/alloc_intrinsic_nontransient_fail.stderr

```diff
@@ -1,8 +1,10 @@
-error: untyped pointers are not allowed in constant
+error: unsupported untyped pointer in constant
   --> $DIR/alloc_intrinsic_nontransient_fail.rs:6:1
    |
 LL | const FOO: *const i32 = foo();
    | ^^^^^^^^^^^^^^^^^^^^^
+   |
+   = note: pointers cast from integers or from `const_allocate` are not allowed in the final value of constants
```

Comment:
This text isn't right. It's not about where the pointer comes from, it's about the type: if this was an `&i32` things would be fine. Also ptrs cast from int *are* allowed, e.g. `const FOO: *const i32 = 0 as *const i32` works fine.

Reply:
Yeah, I was getting a bit confused about this message. oli suggested this on discord: "Only integers cast to pointers or references are supported".

But the following also emits the error: [play](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=7cfd656c3a19efada59cedf88306fc34)

```
struct S;

impl Drop for S {
    fn drop(&mut self) {}
}

const X: *const S = &S;
```

Reply:
Oh right, since the reference turns into a pointer for the constant it isn't allowed.

Reply:
What happens is that we recursively walk the constant in a type-directed way, following the references. Once we are done, we check if there are any pointers left from the constant to some memory that has not been accounted for -- and if that is the case, we bail, since we don't have reliable type information for that memory so we don't know how to intern it.

Reply:
Changed to "only integers cast to pointers or references are supported."

Reply:
I would phrase it as "memory only reachable via raw pointers is not supported"

Reply:
changed.